### ISSUE TITLE
Allow tagged templates to refer to a 'service api' for XML templates

### ISSUE THREAD

**Issue by jbondc** [Microsoft/TypeScript#2759]
This is mainly a rough idea for #296, #2673

The TypeScript service api could expose XML parsing utilities and checking needed by an external service (TAGS: tagged template service) to compile() and check().

Use case:
```javascript
let button = template`templates/ui/button.jsx`
button.set('icon', jsx`<icon class="{props.class}"></icon>`);
button.get('icon').set({class: 'smile'});
```
As you focus into the template string `templates/ui/button.jsx`, TAGS of name 'templates' reads `templates/ui/button.jsx` and shows inline `<div class="button"><icon class="{props.icon.class}"></icon></div>` into your editor. As you unfocus, TAGS  of name 'templates'  saves back to the file. The TAGS of name 'jsx' transforms the template string into an element + provides type checking inside the string.

Tagged templates offer a big domain of possibilities.

Possible api?
```javascript
    export interface TaggedTemplateService {
        compile(text: string, ts: TypeScriptService): TaggedTemplateRootElement
        check(element: TaggedTemplateRootElement): DiagnosticMessageChain
    }
```

**Comment by paulvanbrenk** [comment]
On the page where the JSX spec is discussed, they have a whole paragraph on why template literals are a bad idea. Long story short, your code gets very complex very quickly and is hard to follow. [0]

In general TypeScript always had the philosophy that all JavaScript is valid TypeScript, so our implementation of the JSX support should follow that philosophy and not force users to convert all their code.

[0] http://facebook.github.io/jsx/#why-not-template-literals

**Comment by jbondc** [comment]
That's BOPN (bunch of opinionated nonsense) built around Facebook's implementation .

**Comment by jbondc** [comment]
"It would be necessary to define how JSX behaves within the rest of the ECMAScript grammar within the template literal anyway" - not true, just standardize the apis in the evaluation context, not the grammar.
